### PR TITLE
feat(state): receipts pruning

### PR DESCRIPTION
**Pull request by rkrasiuk**
## Description

Integrates receipts pruning into the new State API.

### CODE REVIEW

**Comment by rakita** [comment]
@rkrasiuk is it okay to go in?

**Comment by rkrasiuk** [comment]
@rakita i still need to resolve conflicts and address some concerns expressed by @joshieDo & @shekhirin

**Comment by rakita** [comment]
Just so we dont forget, I ignored pruning tests on revm_state branch as it was failing: https://github.com/paradigmxyz/reth/commit/8e632f6482b835b73ee34cd2d950d126c57416b0#diff-26fa094d8265997979af6c8880960f5c56bb7f961b9e80c1ed5898682951fa5aR69

File: crates/revm/src/processor.rs [hunk 1]
```diff
@@ -31,14 +33,17 @@ pub struct EVMProcessor<'a> {
     chain_spec: Arc<ChainSpec>,
     evm: EVM<StateDBBox<'a, Error>>,
     stack: InspectorStack,
-    receipts: Vec<Vec<Receipt>>,
+    /// The collection of receipts.
+    receipts: Vec<HashMap<usize, Receipt>>,
```
Comment (shekhirin) [inline crates/revm/src/processor.rs:37]: would appreciate a comment on why `Vec<HashMap<_>>` and not a `Vec<Vec<_>>` as we do in other places

File: crates/revm/src/processor.rs [hunk 2]
```diff
@@ -31,14 +33,17 @@ pub struct EVMProcessor<'a> {
...
     /// First block will be initialized to ZERO
     /// and be set to the block number of first block executed.
     first_block: Option<BlockNumber>,
     /// The maximum known block .
     tip: Option<BlockNumber>,
     /// Pruning configuration.
     prune_modes: PruneModes,
+    /// Memoized address pruning filter.
```
Comment (shekhirin) [inline crates/revm/src/processor.rs:45]: same nit about explanation on what this data structure represents

**Review by shekhirin — APPROVED** [review]
LGTM implementation, a few nits on the comments to make it easier to understand pruning for posterity

File: crates/storage/provider/src/change.rs [hunk 3]
```diff
@@ -349,12 +354,20 @@ impl BundleStateWithReceipts {
...
+            if !receipts.is_empty() {
+                let (_, body_indices) = bodies_cursor
+                    .seek_exact(self.first_block + idx as u64)?
+                    .expect("body indices exist");
+
+                for (tx_idx, tx_num) in body_indices.tx_num_range().enumerate() {
+                    if let Some(receipt) = receipts.remove(&tx_idx) {
+                        receipts_cursor.append(tx_num, receipt)?;
+                    }
+                }
```
Comment (rakita) [inline crates/storage/provider/src/change.rs:370]: Would do this a little bit differently.

You are iterating over blocks `tx_num_range` and then ask `receipts` if there is that number there, so if receipts are empty, you would query hashmap every time.

Would:
```
let first_tx_index = body_indices.first_tx_num();
let mut receipts = receipts.drain().collect::<Vec<_>>();
receipts.sort() // by tx id
for (id,receipt) in receipts {
      receipts_cursor.append(first_tx_index+id,receipt)?;
}
```

**Review by rakita — APPROVED** [review]
lgtm

**Pull request merged**